apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/pluginapi/TransactionInstrumentation.java

```diff
@@ -147,4 +148,41 @@ public static void addCustomContext(@Advice.FieldValue(value = "span", typing =
             }
         }
     }
+
+    public static class SetServiceInfoInstrumentation extends TransactionInstrumentation {
+        public SetServiceInfoInstrumentation() {
+            super(named("setServiceInfo"));
+        }
+
+        public static class AdviceClass {
+            @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+            public static void setServiceInfo(@Advice.FieldValue(value = "span", typing = Assigner.Typing.DYNAMIC) Object transaction,
+                                              @Advice.Argument(0) String serviceName, @Advice.Argument(1) String serviceVersion) {
+                if (transaction instanceof Transaction) {
+                    ((Transaction) transaction).getTraceContext().setServiceName(serviceName);
```

Comment:
I wonder if we should check for `serviceName` being `null`

Reply:
Not sure about this one...
If we allow `null` here, then what happens? If the `service_name` config was manually set, then it would have no effect, right? So this is only a way to use the somewhat useless fallback auto service name?
If so - we can check and reject null values here

Reply:
if `serviceName` is null, but `serviceVersion` is not, `service.name` won't be serialized, but `service.version` will be. I am not sure what the apm-server does in such a case. But I suspect it will drop  transaction specific service configuration and use the global one.

Reply:
I personally think both methods should return early (do no nothing) when `serviceName` is `null`.

Reply:
We could do that centrally by creating a `setServiceInfo(String serviceName, String serviceVersion)` method in `TraceContext`.
Similar to the refactoring in https://github.com/elastic/apm-agent-java/pull/2444, `TraceContext#setServiceInfo` and `TraceContext#setApplicationClassLoader` would be the only methods that set the service name and version.

Reply:
Done!